src/entry.rs

```diff
@@ -0,0 +1,230 @@
+use crate::tree::SmallNode;
+use crate::SgMap;
+
+/// A view into an occupied entry in a `SgMap`.
+/// It is part of the [`Entry`] enum.
+pub struct OccupiedEntry<'a, K: Ord + Default, V: Default, const N: usize> {
+    pub(super) node_idx: usize,
+    pub(super) table: &'a mut SgMap<K, V, N>,
+}
+
+/// A view into a vacant entry in a `SgMap`.
+/// It is part of the [`Entry`] enum.
+pub struct VacantEntry<'a, K: Ord + Default, V: Default, const N: usize> {
+    pub(super) key: K,
+    pub(super) table: &'a mut SgMap<K, V, N>,
+}
+
+impl<'a, K: Ord + Default, V: Default, const N: usize> VacantEntry<'a, K, V, N> {
+    /// Gets a reference to the key that would be used when inserting a value
+    /// through the VacantEntry.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use scapegoat::SgMap;
+    ///
+    /// let mut map = SgMap::<&str, usize, 2>::new();
+    /// assert_eq!(map.entry("poneyland").key(), &"poneyland");
+    /// ```
+    pub fn key(&self) -> &K {
+        &self.key
+    }
+
+    /// Take ownership of the key.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use scapegoat::SgMap;
+    /// use scapegoat::map_types::Entry;
+    ///
+    /// let mut map = SgMap::<&str, usize, 2>::new();
+    ///
+    /// if let Entry::Vacant(v) = map.entry("poneyland") {
+    ///     v.into_key();
+    /// }
+    /// ```
+    pub fn into_key(self) -> K {
+        self.key
+    }
+
+    /// Sets the value of the entry with the `VacantEntry`'s key,
+    /// and returns a mutable reference to it.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use scapegoat::SgMap;
+    /// use scapegoat::map_types::Entry;
+    ///
+    /// let mut map = SgMap::<&str, u32, 2>::new();
+    ///
+    /// if let Entry::Vacant(o) = map.entry("poneyland") {
+    ///     o.insert(37);
+    /// }
+    /// assert_eq!(map["poneyland"], 37);
+    /// ```
+    pub fn insert(self, value: V) -> &'a mut V {
+        todo!()
+    }
```

Comment:
If I use the `SgMap::insert` function here, then the old value is returned. How do I get an exclusive reference to the new value, especially since the key is moved to the map and `Clone` is not implemented by the key?

Reply:
Ah, I see what you mean. You need the index for the newly inserted key-val pair, to use `arena[new_node_idx].get_mut().1` for the exclusive val ref like you've done on line 135.

Even if the key was clone-able, calling `self.table.bst.priv_get(None, &self.key)` (after calling `SgMap::insert`) to determine `new_node_idx` repeats the search/lookup (1 binary search to insert, 1 binary search to get). 2 searches isn't ideal given the "spirit" of the entry APIs (which aim to eliminate redundant lookups). But the key doesn't impl `Clone`, as you've noted, so that's a no-go anyway!

I think the solution is to modify the return signature of `SgTree`'s `priv_balancing_insert` and it's callee `priv_insert` from `-> Option<V>` to `-> (Option<V>, NodeGetHelper<U>)` or similar. Collect the index metadata while inserting, to only need the 1 search. And then use this private insert to implement `VacantEntry`'s `insert`. You'll have to update multiple places in the code (sorry if that's annoying), but the change will make this internal API flexible enough to support your new features and potentially other future stuff.